File: src/p0017/golang/solution.go
var n2s = []string{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}

func letterCombinations(digits string) (result []string) {
	if len(digits) > 0 {
		dfs(digits, 0, "", &result)
	}
	return result
}

func dfs(digits string, i int, s string, result *[]string) {
	if i == len(digits) {
		*result = append(*result, s)
		return
	}
	button := n2s[digits[i]-'0']
	for p := 0; p < len(button); p++ {
		dfs(digits, i+1, s+button[p:p+1], result)
	}
}

File: src/p0017/rust/solution.rs
impl Solution {
    const N2S: &'static [&'static str] = &["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
    pub fn letter_combinations(digits: String) -> Vec<String> {
        let mut result = Vec::new();
        if !digits.is_empty() {
            let s = &mut String::new();
            Self::dfs(digits.as_bytes(), 0, s, &mut result)
        }
        return result;
    }
    fn dfs(digits: &[u8], i: usize, s: &mut String, result: &mut Vec<String>) {
        if i == digits.len() {
            result.push(s.clone());
            return;
        }
        let button = Self::N2S[digits[i] as usize - '0' as usize];
        for x in button.chars() {
            s.push(x);
            Self::dfs(digits, i + 1, s, result);
            s.pop();
        }
    }
}
